wrangler-service/src/main/java/io/cdap/wrangler/service/directive/AbstractDirectiveHandler.java

```diff
@@ -85,16 +100,57 @@ public void destroy() {
       composite.close();
     } catch (IOException e) {
       // If something bad happens here, you might see a a lot of open file handles.
-      LOG.warn("Unable to close the directive registry. You might see increasing number of open file handle.",
-               e.getMessage());
+      LOG.warn("Unable to close the directive registry. You might see increasing number of open file handle.", e);
     }
   }
 
   protected List<Row> executeDirectives(String namespace, List<String> directives,
                                         List<Row> sample) throws DirectiveParseException {
-    Function<TransientStore, ExecutorContext> contextProvider =
-      store -> new ServicePipelineContext(namespace, ExecutorContext.Environment.SERVICE, getContext(), store);
-    return new CommonDirectiveExecutor(contextProvider, composite).executeDirectives(namespace, directives, sample);
+
+    if (directives.isEmpty()) {
+      return sample;
+    }
+
+    GrammarMigrator migrator = new MigrateToV2(directives);
+    String recipe = migrator.migrate();
+
+    // Adds #pragma for UDDs
+    try {
+      Set<String> userDirectives = new HashSet<>();
+      GrammarWalker walker = new GrammarWalker(new RecipeCompiler(), new ConfigDirectiveContext(DirectiveConfig.EMPTY));
+      walker.walk(recipe, (command, tokenGroup) -> {
+        DirectiveInfo info = SystemDirectiveRegistry.INSTANCE.get(namespace, command);
+        if (info == null) {
+          userDirectives.add(command);
+        }
+      });
+      if (!userDirectives.isEmpty()) {
+        directives.add(0, "#pragma load-directives " + String.join(",", userDirectives) + ";");
```

Comment:
Where is this guy used? I see you create a `recipe` above, but I don't see any `directives` usage below. Is it an expected side effect?

Reply:
It is carrying from the old logic that was missed when porting from the old DirectiveHandler to the new WorkspaceHandler.

It is needed for loading UDD.

Reply:
Yes, I understand the intent. Could you point the method this new directive is going to be used? It's not used in this method as far as I can see.

Reply:
In the v2 API this will be used in WorkspaceHandler#specification()

Reply:
Got you. Are there a reason it's added here? Such side effects are very easy to miss while refactoring. Also a very similar code is now present in two places. Can it be a separate method explicitly called from WorkspaceHandler#specification?

Reply:
We need parsing of recipe to find the user defined directive names that has to be appended to #pragma load-directives. This is automatically added only when UDD is being used. Separate method for this will involve parsing it again. Best I can think of to remove it as a side effect is to modify the return type for executeDirectives() to return both List<Row> and pragma string/udd names and let the caller add it to directives. @chtyim , thoughts?